File: src/levels/L05-adapter/ts/facade.ts
class Cashier {
  constructor(
    private payments: Payments,
    private barista: Barista,
  ) {}

  // region: Cashier#charge
  charge(order: Order) {
    this.payments.charge(order.total)
    this.barista.prepare(order)
  }
  // endregion
}

// Una fachada simplifica un subsistema, pero no traduce: sigue pidiendo charge().
// region: Payments#charge
class Payments {
  constructor(private sdk: PaymentGateway) {}
  charge(total: number) { this.sdk.charge(total) }
}
// endregion

File: src/levels/L05-adapter/facade.rb
class Cashier
  def initialize(payments, barista)
    @payments = payments
    @barista = barista
  end

  # region: Cashier#charge
  def charge(order)
    @payments.charge(order.total)
    @barista.prepare(order)
  end
  # endregion
end

# Una fachada simplifica un subsistema, pero no traduce: sigue pidiendo #charge.
# region: Payments#charge
class Payments
  def initialize(sdk) = @sdk = sdk
  def charge(total) = @sdk.charge(total)
end
# endregion
